runtime/runtime/src/congestion_control.rs

```diff
@@ -0,0 +1,372 @@
+            // No limits on receipts that stay on the same shard. Backpressure
+            // wouldn't help, the receipt takes the same memory if buffered or
+            // in the delayed receipts queue.
+            self.outgoing_receipts.push(receipt);
+            return Ok(());
+        }
+        match Self::try_forward(
+            receipt,
+            shard,
+            &mut self.outgoing_limit,
+            self.outgoing_receipts,
+            apply_state,
+        )? {
+            ReceiptForwarding::Forwarded => (),
+            ReceiptForwarding::NotForwarded(receipt) => {
+                self.buffer_receipt(&receipt, state_update, shard, &apply_state.config)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Forward a receipt if possible and return whether it was forwarded or
+    /// not.
+    ///
+    /// This does not take `&mut self` as first argument to make lifetime
+    /// management easier. Instead it takes exactly the fields it requires,
+    /// namely `outgoing_limit` and `outgoing_receipt`.
+    fn try_forward(
+        receipt: Receipt,
+        shard: ShardId,
+        outgoing_limit: &mut HashMap<ShardId, Gas>,
+        outgoing_receipts: &mut Vec<Receipt>,
+        apply_state: &ApplyState,
+    ) -> Result<ReceiptForwarding, RuntimeError> {
+        // Default case set to `Gas::MAX`: If no outgoing limit was defined for the receiving
+        // shard, this usually just means the feature is not enabled. Or, it
+        // could be a special case during resharding events. Or even a bug. In
+        // any case, if we cannot know a limit, treating it as literally "no
+        // limit" is the safest approach to ensure availability.
+        let forward_limit = outgoing_limit.entry(shard).or_insert(Gas::MAX);
+        let gas_to_forward = receipt_congestion_gas(&receipt, &apply_state.config)?;
+        if *forward_limit > gas_to_forward {
+            outgoing_receipts.push(receipt);
+            // underflow impossible: checked forward_limit > gas_to_forward above
+            *forward_limit -= gas_to_forward;
+            Ok(ReceiptForwarding::Forwarded)
+        } else {
+            Ok(ReceiptForwarding::NotForwarded(receipt))
+        }
+    }
+
+    /// Put a receipt in the outgoing receipt buffer of a shard.
+    fn buffer_receipt(
+        &mut self,
+        receipt: &Receipt,
+        state_update: &mut TrieUpdate,
+        shard: u64,
+        config: &RuntimeConfig,
+    ) -> Result<(), RuntimeError> {
+        let bytes = receipt_size(&receipt)?;
+        let gas = receipt_congestion_gas(&receipt, config)?;
+        self.congestion_info.add_receipt_bytes(bytes as u64)?;
+        self.congestion_info.add_buffered_receipt_gas(gas)?;
+        self.outgoing_buffers.to_shard(shard).push(state_update, &receipt)?;
+        Ok(())
+    }
+}
+
+fn receipt_congestion_gas(
+    receipt: &Receipt,
+    config: &RuntimeConfig,
+) -> Result<Gas, IntegerOverflowError> {
```

Comment:
I'm not qualified to review this, perhaps someone from runtime can have a look?

Reply:
@akhi3030 can you please suggest someone to review `receipt_congestion_gas` in this PR?

It is a method that computes how much work we will do at most for a receipt, measured in gas. Naively, it should be just attached gas. But there is some prepaid gas as well, which is paid in the previous block but we burn it this block, so the work must be accounted for in this block. Plus we need to look details of different receipt types.

So, we are looking for someone with good knowledge / willingness to drill deep about how gas spending is done in multiple steps and how it relates to actual gas burning. I can also help someone building that knowledge, if necessary.

Reply:
@jakmeier: let me discuss and get back to you.